File: src/aoc2021/day17.clj
(ns aoc2021.day17
  (:require [clojure.java.io :as io]))

(defn parse
  [data]
  (let [[left right bottom top] (map read-string (re-seq #"-?\d+" data))]
    {:left left :right right :top top :bottom bottom}))

(def sample-input
  (-> "../resources/day17_ex.txt" io/resource io/reader slurp parse))

(def puzzle-input
  (-> "../resources/day17.txt" io/resource io/reader slurp parse))

(defn step
  [{:keys [y x-velocity y-velocity] :as probe}]
  (-> probe
      (update :x + x-velocity)
      (update :y + y-velocity)
      (update :max-y max y)
      (update :x-velocity (cond
                            (pos? x-velocity)  dec
                            (zero? x-velocity) identity
                            (neg? x-velocity)  inc))
      (update :y-velocity dec)))

(defn hit-target?
  [{:keys [left right top bottom]} {:keys [x y] :as probe}]
  (cond
    (and (<= left x right) (<= bottom y top)) probe
    (< y bottom)                              false
    (or (< x left) (> y top))                 nil
    :else                                     false))

(defn fire-probe
  [target x-velocity y-velocity]
  (when-not (zero? x-velocity)
    (->> {:x 0 :y 0 :x-velocity x-velocity :y-velocity y-velocity :max-y 0}
         (iterate step)
         (keep (partial hit-target? target))
         first)))

;; to hit the target, x-velocity must be positive but
;; cannot be greater than target's right edge
;; y-velocity cannot be less than target's bottom edge, but what's the upper bound?
;; trial and error (and the part 2 example) shows that it's the opposite value
;; of the target's bottom edge

(defn solve
  [{:keys [right bottom] :as target}]
  ((juxt #(:max-y (apply max-key :max-y %))
         count)
   (for [xv (range 1 (inc right))
         yv (range bottom (inc (- bottom)))
         :let [result (fire-probe target xv yv)]
         :when result]
     {:max-y (:max-y result) :init-xv xv :init-yv yv})))

File: py/day17.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

import re

def parse(path):
    with open(path) as f:
        left, right, bottom, top = [int(d) for d in re.findall(r"-?\d+", f.readline())]
        return {'left': left, 'right': right, 'bottom': bottom, 'top': top}

sample_input = parse('../resources/day17_ex.txt')
puzzle_input = parse('../resources/day17.txt')

class Probe:
    def __init__(self, xv, yv):
        self.x       = 0
        self.y       = 0
        self.xv      = xv
        self.yv      = yv
        self.max_y   = 0
        self.success = None

    def step(self):
        self.x     += self.xv
        self.y     += self.yv
        self.max_y = max(self.y, self.max_y)
        self.yv    -= 1
        if self.xv > 0:
            self.xv -= 1

    def launch(self, left, right, bottom, top):
        while self.success is None:
            self.step()
            if (left <= self.x <= right) and (bottom <= self.y <= top):
                self.success = True
            elif (self.y < bottom):
                self.success = False
            elif (self.x < left) or (self.y > top):
                pass
            else:
                self.success = False

def solve(left, right, bottom, top):
    max_y, count = 0, 0
    for xv in range(1, right + 1):
        for yv in range(bottom, (bottom * -1) + 1):
            probe = Probe(xv, yv)
            probe.launch(left, right, bottom, top)
            if probe.success == True:
                max_y = max(max_y, probe.max_y)
                count += 1
    return (max_y, count)

print("Day 17\n"
      "[Part 1] Example: {example[0]}, Puzzle: {solution[0]}\n"
      "[Part 2] Example: {example[1]}, Puzzle: {solution[1]}".format
      (example=solve(**sample_input), solution=solve(**puzzle_input)))
